File: app/src/main/java/com/vikingsen/cheesedemo/util/SchedulerProvider.java
package com.vikingsen.cheesedemo.util;


import io.reactivex.Scheduler;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public interface SchedulerProvider {
    Scheduler computation();
    Scheduler io();
    Scheduler ui();

    class AppSchedulerProvider implements SchedulerProvider {

        @Override
        public Scheduler computation() {
            return Schedulers.computation();
        }

        @Override
        public Scheduler io() {
            return Schedulers.io();
        }

        @Override
        public Scheduler ui() {
            return AndroidSchedulers.mainThread();
        }
    }

    class TrampolineSchedulerProvider implements SchedulerProvider {

        @Override
        public Scheduler computation() {
            return Schedulers.trampoline();
        }

        @Override
        public Scheduler io() {
            return Schedulers.trampoline();
        }

        @Override
        public Scheduler ui() {
            return Schedulers.trampoline();
        }
    }
}

File: app/src/main/java/com/vikingsen/cheesedemo/util/SchedulerProvider.kt
package com.vikingsen.cheesedemo.util


import io.reactivex.Scheduler
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

interface SchedulerProvider {
    fun computation(): Scheduler
    fun io(): Scheduler
    fun ui(): Scheduler

    class AppSchedulerProvider : SchedulerProvider {

        override fun computation(): Scheduler {
            return Schedulers.computation()
        }

        override fun io(): Scheduler {
            return Schedulers.io()
        }

        override fun ui(): Scheduler {
            return AndroidSchedulers.mainThread()
        }
    }

    class TrampolineSchedulerProvider : SchedulerProvider {

        override fun computation(): Scheduler {
            return Schedulers.trampoline()
        }

        override fun io(): Scheduler {
            return Schedulers.trampoline()
        }

        override fun ui(): Scheduler {
            return Schedulers.trampoline()
        }
    }
}
